docs/src/main/asciidoc/writing-native-applications-tips.adoc

```diff
@@ -344,6 +344,231 @@ com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfac
 Solving this issue requires adding the `-H:DynamicProxyConfigurationResources=<comma-separated-config-resources>` option and to provide a dynamic proxy configuration file.
 You can find all the information about the format of this file in https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/DynamicProxy.md#manual-configuration[the GraalVM documentation].
 
+[[modularity-benefits]]
+=== Modularity Benefits
+
+As already explained in sections above,
+during native image build time GraalVM analyses the application's call tree and generates a universe of all the code it needs.
+Having a modular codebase is key to avoiding problems with unused or optional parts of your application,
+while at the same time reducing both native image build times and native executable size.
+In this section you will learn about the details behind the benefits of modularity for native applications.
+
+When code is not modular enough, generated native executables can end up with more code than what the user needs.
+If a feature is not used and the code gets compiled into the native executable,
+this is both a waste of native compilation time and native executable disk space.
+Even more problems arise when third party libraries or specialized API subsystems are used which cause native compilation or runtime errors,
+and their use is not modularised enough.
+A recent problem can be found in the jaxb library,
+which is capable of deserializing XML files containing images using Java’s AWT APIs.
+The vast majority of Quarkus XML users don’t need to deserialize images,
+so there shouldn’t be a need for users applications to include Java AWT code,
+unless they would specifically configure Quarkus to add the jaxb AWT code to the native executable.
+However, because jaxb code that uses AWT is in the same jar as the rest of the XML parsing code,
+achieving this separation was rather complex and required the use of Java bytecode substitutions to get around it.
+These substitutions are hard to maintain and can easily break, hence they are really the last line of defence.
+
+A modular codebase is the best way to avoid these kind of issues.
+Taking the jaxb/AWT problem above,
+if the jaxb code that dealt with images was in a separate module or jar (e.g. `jaxb-images`),
+then Quarkus could choose not to include that module unless the user specifically requested the need to serialize/deserialize XML files containing images at build time.
+
+An added benefit of modular applications can make the universe of code that gets into the native executable smaller.
+The smaller the universe, the faster it native image builds will be and the smaller the native executable produced.
+
+The key takeaway point here is the following:
+Keeping optional features, particularly those that depend on third party libraries or API subsystems with a big footprint,
+in separate modules is the best solution.
```

Comment:
This might be worth putting in an asciidoc block.

Reply:
I'm wrapping this in a tip block. Is that what you had in mind?

Reply:
Yes. Thank you